File: C++/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        if (s.size() % 2 == 1) return false;
        for (char a : s) {
            if (a == '(') st.push(')');
            else if (a == '[') st.push(']');
            else if (a == '{') st.push('}');
            else if (st.empty() || st.top() != a) return false;
            else st.pop();
        }
        return st.empty();
    }
};

File: Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        st = []
        map_ = {'{': '}', '[': ']', '(':')'}
        for a in s:
            if a in ['{', '[', '(']:
                st.append(map_[a])
            elif len(st) == 0 or st[-1] != a:
                return False
            else:
                st.pop(-1)
        return len(st) == 0